### PR TITLE
[Enhancement] Support add multi range partition manually to expression partition table

### PR DESCRIPTION
**Pull request by HangyuanLiu**
## Why I'm doing:

- Currently, PartitionDesc does not have a unified Analyze logic, which is scattered in each execution flow of the code. However, partition desc is used in many places, such as the create table, CTAS, and Materialized View, which often results in inconsistent performance. New in this PR is PartitionDescAnalyzer, but this PR only sets up a framework that will require ongoing code migration and refactoring in the future.   

- The PR supports manually adding partitions to the auto partition table

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by nshangyiming — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionDescAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    public static void analyze(PartitionDesc partitionDesc) {
+        if (partitionDesc instanceof SingleRangePartitionDesc) {
+            analyzeSingleRangePartitionDesc((SingleRangePartitionDesc) partitionDesc);
+        } else if (partitionDesc instanceof MultiRangePartitionDesc) {
+            analyzeMultiRangePartitionDesc((MultiRangePartitionDesc) partitionDesc);
+        }
+    }
+
+    public static void analyzeSingleRangePartitionDesc(SingleRangePartitionDesc singleRangePartitionDesc) {
+    }
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionDescAnalyzer.java:54]: will add analzye step in the future?

**Review by packy92 — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionDescAnalyzer.java (hunk 1 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionDescAnalyzer.java:54]: yes

**Review by mervinsr — APPROVED** [review]

**Review by meegoo — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 2]
```diff
@@ -311,6 +311,21 @@ public enum ErrorCode {
     ERR_TOO_MANY_ERROR_ROWS(5606, new byte[] {'2', '2', '0', '0', '0'},
             "Current error rows: %d is more than max error num: %d. Check the 'TrackingSQL' field for detailed information"),
 
+    /**
+     * 5700 - 5799: Partition
+     */
+    ERR_ADD_PARTITION_WITH_ERROR_PARTITION_TYPE(5700, new byte[] {'5', '8', '0', '3', '0'},
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:317]: If it's in the SQL parse stage, it's better to use 42601 as the SQLState.

**Review by nshangyiming — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java (hunk 2 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:317]: not parse stage

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 3]
```diff
@@ -311,6 +311,21 @@ public enum ErrorCode {
...
+            "Cannot add a Range partition to a table that partition type is not of Range"),
+
+    ERR_ADD_PARTITION_WITH_ERROR_STEP_LENGTH(5701, new byte[] {'5', '8', '0', '3', '0'},
+            "Step length [%d] in the operation is not equal to the partition step length [%d] stored in the table"),
+
+    ERR_MULTI_PARTITION_COLUMN_NOT_SUPPORT_ADD_MULTI_RANGE(5702,  new byte[] {'5', '8', '0', '3', '0'},
+            "Can't add multi range partition to multi partition column table"),
```
[the remark this answers is not in the document]
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:324]: There is no need for this, it is meaningless information, users can easily know the table name

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 4]
```diff
@@ -311,6 +311,21 @@ public enum ErrorCode {
     ERR_TOO_MANY_ERROR_ROWS(5606, new byte[] {'2', '2', '0', '0', '0'},
             "Current error rows: %d is more than max error num: %d. Check the 'TrackingSQL' field for detailed information"),
 
+    /**
+     * 5700 - 5799: Partition
+     */
+    ERR_ADD_PARTITION_WITH_ERROR_PARTITION_TYPE(5700, new byte[] {'5', '8', '0', '3', '0'},
+            "Cannot add a Range partition to a table that partition type is not of Range"),
```
[the remark this answers is not in the document]
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:318]: There is no need for this, it is meaningless information, users can easily know the partition type

**Review by jaogoy — DISMISSED** [review]

**Pull request merged**